### PR TITLE
chore(tracing): add span.kind tag 

### PR DESCRIPTION
**Pull request by wconti27**
## Description
Changed general span attributes. Added a span.kind attribute that will be populated for relevant integration spans with one of client, producer, consumer, or server. Spans without the span.kind attribute will be considered internal. 

** This branch was created by branching off of the branch adding component attribute within the Python tracer. Once the component tag change PR is merged, this PR will be changed to be merged into 1.x **

Also providing an Excel file with all integrations, labeled with what integrations were changed, what tests were changed, etc
Document can be found [here](https://docs.google.com/spreadsheets/d/1F8yqzdi9AncVRy--k6Evw235PAudRZ4lYu5HSojXM8A/edit?usp=sharing).

## Motivation
Part of a larger effort to make tag naming more consistent among Datadog tracing libraries. Component is a required attribute for spans created from an integration.

## Design 
Span.Kind tag was set directly within each integrations after initially starting to trace a span if span was tracing a relevant span.kind operation.

Drawbacks of the change are that this PR is huuuuuuuuge. Sorry. 

## Testing strategy
Tested span.kind field by adding assertions on span.kind field to unit/integration tests. Also fixed snapshots to contain newly added span.kind field. 

Newly added general tags are also being tested in the system testing repo as well with this PR (component, span.kind, language, process_id): [add_meta_tag_testing ](https://github.com/DataDog/system-tests/pull/645)

## Relevant issue(s)
Relevant related issues: 
- [Add component tag within Python tracer ](https://github.com/DataDog/dd-trace-py/pull/4524)
- [Add span.kind tag within Python tracer ](https://github.com/DataDog/dd-trace-py/tree/AIT-5038-add-span-kind-tag)
- [Add language tag within Python tracer ](https://github.com/DataDog/dd-trace-py/pull/4535)
- [Change system.pid tag to process_id within Python tracer ](https://github.com/DataDog/dd-trace-py/pull/4534)

## Testing strategy
? Detailed above

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/constants.py [hunk 1]
```diff
@@ -50,6 +50,14 @@
 
 PID = "process_id"
 
+SPAN_KIND = "span.kind"
+
+# Types of span.kind
+SPAN_CLIENT = "client"
+SPAN_SERVER = "server"
```
Comment (mabdinur) [inline ddtrace/constants.py:57]: Do we expect ddtrace users to use these constants? If these constants will only used in integrations we should define these values in an internal module (ex: ddtrace/internal/constants.py).

Suggestion: We can create a SpanKind class that defines these constants. This better aligns with how we handle [spantypes](https://github.com/DataDog/dd-trace-py/blob/v1.7.2/ddtrace/ext/__init__.py):

ex:
```
Class SpanKind:
    SPAN_CLIENT = "client"
    SPAN_SERVER = "server"
    SPAN_PRODUCER = "producer"
    SPAN_CONSUMER = "consumer"
```

We can define SpanKind in a public or internal module depending on the expected usage.

Comment (wconti27) [reply ddtrace/constants.py:57]: Though these constants are mainly for our internal use, it's not out of the question IMO that a user would want to utilize these constants when manually instrumenting their own code. I like the class idea!

Comment (wconti27) [reply ddtrace/constants.py]: Done. Added the SpanKind class like SpanTypes

File: ddtrace/contrib/aiohttp/patch.py [hunk 2]
```diff
@@ -73,6 +87,9 @@ async def _traced_clientsession_request(aiohttp, pin, func, instance, args, kwar
         # set component tag equal to name of integration
         span.set_tag_str("component", config.aiohttp_client.integration_name)
 
+        # set span.kind tag equal to type of request
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:91]:
```suggestion
        span.set_tag_str(SPAN_KIND, SpanKind.CLIENT)
```

File: ddtrace/contrib/aiohttp/patch.py [hunk 3]
```diff
@@ -45,13 +47,25 @@ def __init__(self, obj, pin):
...
     async def _create_connection(self, req, *args, **kwargs):
         pin = Pin.get_from(self)
-        with pin.tracer.trace("%s._create_connection" % self.__class__.__name__):
+        with pin.tracer.trace("%s._create_connection" % self.__class__.__name__) as span:
+            # set component tag equal to name of integration
+            span.set_tag("component", config.aiohttp.integration_name)
+
+            # set span.kind tag equal to type of request
+            span.set_tag(SPAN_KIND, SpanKind.SERVER)
+
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:68]:
```suggestion
```

I don't think we want these to have span kind ?

File: ddtrace/contrib/aiohttp/patch.py [hunk 4]
```diff
@@ -45,13 +47,25 @@ def __init__(self, obj, pin):
...
     async def connect(self, req, *args, **kwargs):
         pin = Pin.get_from(self)
-        with pin.tracer.trace("%s.connect" % self.__class__.__name__):
+        with pin.tracer.trace("%s.connect" % self.__class__.__name__) as span:
+            # set component tag equal to name of integration
+            span.set_tag("component", config.aiohttp.integration_name)
+
+            # set span.kind tag equal to type of request
+            span.set_tag(SPAN_KIND, SpanKind.SERVER)
+
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:56]:
```suggestion
```

I am not sure if this should have span kind

File: ddtrace/contrib/aiohttp/patch.py (hunk 2 above)
Comment (brettlangdon) [reply ddtrace/contrib/aiohttp/patch.py:91]: everything in this file is for the client, and everything in `middlewares` is for the server (as far as I can tell)

File: ddtrace/contrib/asgi/middleware.py [hunk 5]
```diff
@@ -123,6 +125,9 @@ async def __call__(self, scope, receive, send):
         # set component tag equal to name of integration
         span.set_tag_str("component", self.integration_config.integration_name)
 
+        # set span.kind to the type of request being performed
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
Comment (brettlangdon) [inline ddtrace/contrib/asgi/middleware.py:129]: if we set this here, then we probably don't need to set it in the span modifier for any integration which uses this as it's base.

Same goes for component tag since we are using the passed in integration_config (I am not 100% confident on this component part, but seems like it?)

File: ddtrace/contrib/flask/patch.py [hunk 6]
```diff
@@ -135,6 +137,9 @@ def _request_span_modifier(self, span, environ):
 
         span.set_tag_str(FLASK_VERSION, flask_version_str)
 
+        # set span.kind to the type of request being performed
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:141]: I don't think we need this is WSGI does this for us?

**Review by brettlangdon — COMMENTED** [review]
sorry I didn't get 100% of the way through the file, but thought I'd leave the comments I do have so far (I made it up to Flask)

File: ddtrace/contrib/aiohttp/patch.py (hunk 2 above)
Comment (wconti27) [reply ddtrace/contrib/aiohttp/patch.py]: Yup, good catch, fixed

File: ddtrace/contrib/aiohttp/patch.py (hunk 4 above)
Comment (wconti27) [reply ddtrace/contrib/aiohttp/patch.py]: Agreed, removed any span kind tagging unless the span is a request span.

File: ddtrace/contrib/flask/patch.py (hunk 6 above)
Comment (wconti27) [reply ddtrace/contrib/flask/patch.py]: Yup, I removed this

File: ddtrace/contrib/asgi/middleware.py (hunk 5 above)
Comment (wconti27) [reply ddtrace/contrib/asgi/middleware.py:129]: Yeah you're correct, Django and Starlette use this class for middleware spans. Currently we are already using this class for setting the component and span.kind tags. None of that work is happening in the patches for middleware spans specifically.

File: ddtrace/contrib/requests/connection.py [hunk 7]
```diff
@@ -82,6 +84,9 @@ def _wrap_send(func, instance, args, kwargs):
         # set component tag equal to name of integration
         span.set_tag_str("component", config.requests.integration_name)
 
+        # set span.kind to the type of operation being performed
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
Comment (mabdinur) [inline ddtrace/contrib/requests/connection.py:88]: Client?

File: ddtrace/contrib/urllib3/patch.py [hunk 8]
```diff
@@ -99,6 +101,9 @@ def _wrap_urlopen(func, instance, args, kwargs):
         # set component tag equal to name of integration
         span.set_tag_str("component", config.urllib3.integration_name)
 
+        # set span.kind to the type of operation being performed
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
Comment (mabdinur) [inline ddtrace/contrib/urllib3/patch.py:105]: Client?

File: riotfile.py [hunk 9]
```diff
@@ -156,22 +156,22 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+        #         "mypy": latest,
+        #         "envier": latest,
+        #         "types-attrs": latest,
+        #         "types-docutils": latest,
+        #         "types-protobuf": latest,
+        #         "types-PyYAML": latest,
+        #         "types-setuptools": latest,
+        #         "types-six": latest,
+        #     },
+        # ),
```
Comment (mabdinur) [inline riotfile.py:174]: We should revert this change. This PR should pass mypy checks

File: riotfile.py [hunk 10]
```diff
@@ -186,16 +186,16 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+        # Venv(
+        #     pys=["3"],
+        #     pkgs={"slotscheck": latest},
+        #     venvs=[
+        #         Venv(
+        #             name="slotscheck",
+        #             command="python -m slotscheck -v {cmdargs}",
+        #         ),
+        #     ],
+        # ),
```
Comment (mabdinur) [inline riotfile.py:198]: revert

**Review by mabdinur — COMMENTED** [review]
It is always easier to add a tag than to remove it or update its value (Updating tags is disruptive to users. It can break metrics and facets). We should be very cautious about which spans have the kind tag and which value is set, especially in helper functions that create spans for different operations.

File: ddtrace/contrib/fastapi/patch.py [hunk 11]
```diff
@@ -63,7 +63,10 @@ async def traced_serialize_response(wrapped, instance, args, kwargs):
     if not pin or not pin.enabled:
         return await wrapped(*args, **kwargs)
 
-    with pin.tracer.trace("fastapi.serialize_response"):
+    with pin.tracer.trace("fastapi.serialize_response") as span:
+        # set component tag equal to name of integration
+        span.set_tag_str("component", config.fastapi.integration_name)
+
         return await wrapped(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/fastapi/patch.py:70]: Asked this same question to @brettlangdon awhile ago when this PR was made and he confirmed that it should be internal. I have it noted in my personal Excel notes for this PR

File: ddtrace/contrib/httplib/patch.py [hunk 12]
```diff
@@ -79,6 +81,9 @@ def _wrap_request(func, instance, args, kwargs):
         # set component tag equal to name of integration
         span.set_tag_str("component", config.httplib.integration_name)
 
+        # set span.kind to the type of operation being performed
+        span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/httplib/patch.py]: Correct, fixed this for all http clients that were miscorrectly labeled.

File: ddtrace/contrib/httplib/patch.py [hunk 13]
```diff
@@ -122,6 +127,9 @@ def _wrap_putrequest(func, instance, args, kwargs):
             # set component tag equal to name of integration
             span.set_tag_str("component", config.httplib.integration_name)
 
+            # set span.kind to the type of operation being performed
+            span.set_tag_str(SPAN_KIND, SpanKind.SERVER)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/httplib/patch.py]: resolved

File: riotfile.py (hunk 9 above)
Comment (wconti27) [reply riotfile.py]: Reverted this change, didn't mean for this to sneak into the PR

File: riotfile.py (hunk 10 above)
Comment (wconti27) [reply riotfile.py]: same as above, reverted

File: ddtrace/ext/__init__.py [hunk 14]
```diff
@@ -12,3 +12,10 @@ class SpanTypes(object):
     TEST = "test"
     WEB = "web"
     WORKER = "worker"
+
+
+class SpanKind(object):
+    CLIENT = "client"
+    SERVER = "server"
+    PRODUCER = "producer"
+    CONSUMER = "consumer"
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/ext/__init__.py:21]: Considering that we don't tag `span.kind` for internal spans, is there any use in adding a constant? Though I guess we can add it to match OTel. @brettlangdon any thoughts?

**Review by majorgreys — CHANGES_REQUESTED** [review]
Following recent discussions on this, I would ask that the type of change here is a chore as no public interface changes are being made, nor are users impacted. If a user can opt in or out of this new behavior, then we'd want it to be flagged as such. But as it is, it is a change to the internal data model.

File: ddtrace/contrib/aiohttp/patch.py [hunk 15]
```diff
@@ -46,13 +48,19 @@ def __init__(self, obj, pin):
 
     async def connect(self, req, *args, **kwargs):
         pin = Pin.get_from(self)
-        with pin.tracer.trace("%s.connect" % self.__class__.__name__):
+        with pin.tracer.trace("%s.connect" % self.__class__.__name__) as span:
+            # set component tag equal to name of integration
+            span.set_tag("component", config.aiohttp.integration_name)
```
Comment (majorgreys) [inline ddtrace/contrib/aiohttp/patch.py:53]: Can the additions of component tag be pulled out?

File: ddtrace/contrib/aiohttp/patch.py [hunk 16]
```diff
@@ -46,13 +48,19 @@ def __init__(self, obj, pin):
...
+
             result = await self.__wrapped__.connect(req, *args, **kwargs)
             return result
 
     async def _create_connection(self, req, *args, **kwargs):
         pin = Pin.get_from(self)
-        with pin.tracer.trace("%s._create_connection" % self.__class__.__name__):
+        with pin.tracer.trace("%s._create_connection" % self.__class__.__name__) as span:
+            # set component tag equal to name of integration
+            span.set_tag("component", config.aiohttp.integration_name)
```
Comment (majorgreys) [inline ddtrace/contrib/aiohttp/patch.py:62]: Same as above

File: ddtrace/contrib/aiohttp/patch.py (hunk 15 above)
Comment (wconti27) [reply ddtrace/contrib/aiohttp/patch.py:53]: yea will add this to another PR

**Review by majorgreys — DISMISSED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]

**Review by ZStriker19 — DISMISSED** [review]

**Review by emmettbutler — DISMISSED** [review]

**Review by mabdinur — APPROVED** [review]

**Review by ZStriker19 — DISMISSED** [review]

File: ddtrace/contrib/flask/patch.py [hunk 17]
```diff
@@ -636,6 +638,9 @@ def _set_request_tags(span):
 
         span.set_tag_str(COMPONENT, config.flask.integration_name)
 
+        if span.name.split(".")[-1] == "request":
```
Comment (majorgreys) [inline ddtrace/contrib/flask/patch.py:641]: Would this not always be true for this function?

**Review by majorgreys — COMMENTED** [review]
one question with the flask change

File: ddtrace/contrib/flask/patch.py (hunk 17 above)
Comment (majorgreys) [inline ddtrace/contrib/flask/patch.py:641]: not-blocking: We should not be calling this function when the span is not the flask request. This is a result of how the wsgi middleware is used to update the span resource.

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]
